fe1-web/network/NetworkConnection.ts

```diff
@@ -0,0 +1,155 @@
+import { w3cwebsocket as W3CWebSocket } from 'websocket';
+import {
+  JsonRpcRequest, JsonRpcResponse, ProtocolError, UNDEFINED_ID,
+} from 'model/network';
+import { NetworkError } from './NetworkError';
+
+type ResponseHandler = (message: JsonRpcResponse) => void;
+
+const DEFAULT_RESPONSE_HANDLER: ResponseHandler = (message: JsonRpcResponse) => {
+  if (message.error === undefined) {
+    // default handler does nothing which is useful for some messages
+  } else {
+    console.error(
+      'A negative network error was received:\n'
+      + `\t- error code : ${message.error.code}`
+      + `\t- description : ${message.error.description}`,
+    );
+  }
+};
+
+const WEBSOCKET_READYSTATE_INTERVAL_MS = 10;
+const WEBSOCKET_READYSTATE_MAX_ATTEMPTS = 100;
+
+const WEBSOCKET_MESSAGE_TIMEOUT_MS = 10000; // 10 seconds max round-trip time
+
+interface PendingResponse {
+  promise: Promise<JsonRpcResponse>,
+  resolvePromise: (value: JsonRpcResponse) => void,
+  rejectPromise: () => void,
+  timeoutId: NodeJS.Timeout,
+}
+
+export class NetworkConnection {
+  private readonly ws: any;
```

Comment:
if the websocket disconnects, is there any retry-mechanism to bring it back up alive? 
If not, the first network crash will render the application unusable.